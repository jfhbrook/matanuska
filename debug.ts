import { Attributes, context, Context, Span, SpanOptions, trace } from '@opentelemetry/api';

import { Tree } from './ast';
import { Chunk } from './bytecode/chunk';
import { Runtime } from './runtime';

//#if _MATBAS_BUILD == 'debug'
import { parseBoolEnv } from './env';
import { formatter } from './format';
//#if _DEBUG_SHOW_CHUNK
import { disassemble } from './bytecode/disassembler';
//#endif
//#if _DEBUG_TRACE_RUNTIME
import { disassembleInstruction } from './bytecode/disassembler';
//#endif
//#endif

let NO_TRACE = true;

/**
 * Show a parse tree.
 *
 * @param tree The tree to show.
 */
let showTree = function showTree(_tree: Tree): void {
  return;
};

/**
 * Show a compiled chunk.
 *
 * @param chunk The chunk to show.
 */
let showChunk = function showChunk(_chunk: Chunk): void {
  return;
};

/**
 * Log the start of execution tracing.
 */
let startTraceExec = function startTraceExec(): void {};

/**
 * Trace a step in execution.
 * @param rt The runtime.
 */
let traceExec = function traceExec(_rt: Runtime): void {};

const tracer = trace.getTracer('main');

// Very similar to OpenTelemetry's implementation of startActiveSpan, except
// the callback function closes the span and attaches annotations in error
// cases.
//
// This is currently defined regardless of whether or not we're in debug
// mode, so that calling startSpan without a guard won't crash the program.
// In practice, guards are placed at the call site.
//
// See: https://github.com/open-telemetry/opentelemetry-js/blob/main/api/src/trace/NoopTracer.ts#L56-L100
function startSpan<F extends (span: Span) => ReturnType<F>>(
  name: string,
  fn: F
): ReturnType<F>;
function startSpan<F extends (span: Span) => ReturnType<F>>(
  name: string,
  opts: SpanOptions | undefined,
  fn: F
): ReturnType<F>;
function startSpan<F extends (span: Span) => ReturnType<F>>(
  name: string,
  opts: SpanOptions | undefined,
  ctx: Context | undefined,
  fn: F
): ReturnType<F>;
function startSpan<F extends (span: Span) => ReturnType<F>>(
  name: string,
  arg2?: F | SpanOptions,
  arg3?: F | Context,
  arg4?: F
): ReturnType<F> | undefined {
  let opts: SpanOptions | undefined;
  let ctx: Context | undefined;
  let fn: F;
  if (arguments.length < 2) {
    return;
  } else if (arguments.length === 2) {
    fn = arg2 as F;
  } else if (arguments.length === 3) {
    opts = arg2 as SpanOptions | undefined;
    fn = arg3 as F;
  } else {
    opts = arg2 as SpanOptions | undefined;
    ctx = arg3 as Context | undefined;
    fn = arg4 as F;
  }
  const wrapped = (span: Span) => {
    try {
      return fn(span);
    } catch (err) {
      span.recordException(err);
  throw err;
    } finally {
      span.end();
    }
  }

  const parentContext = ctx ?? context.active();
  const span = tracer.startSpan(name, opts, parentContext);
  const contextWithSpanSet = trace.setSpan(parentContext, span);
  return context.with(contextWithSpanSet, wrapped, undefined, span);
}

// A convenience function for adding events when you don't have the span
// immediately on-hand. Like startSpan, this is not hidden behind jscc and
// should instead be conditionally imported/called at the site of use.
function addEvent(message: string, attributes: Attributes = {}): void {
  const span = trace.getActiveSpan();
  if (span) {
    span.addEvent(message, attributes);
  }
}

//#if _MATBAS_BUILD == 'debug'
NO_TRACE = parseBoolEnv(process.env.NO_TRACE);

//#if _DEBUG_SHOW_TREE
if (!NO_TRACE) {
  showTree = function showTree(tree: Tree): void {
    console.log('=== Parse Tree: ===');
    console.log(formatter.format(tree));
  };
}
//#endif

//#if _DEBUG_SHOW_CHUNK
if (!NO_TRACE) {
  showChunk = function showChunk(chunk: Chunk): void {
    console.log(disassemble(chunk));
  };
}
//#endif

//#if _DEBUG_TRACE_RUNTIME
if (!NO_TRACE) {
  startTraceExec = function startTraceExec(): void {
    console.log('=== Execution Trace: ===');
  };

  traceExec = function traceExec(rt: Runtime): void {
    console.log('> stack:', formatter.format(rt.stack));
    console.log('>', disassembleInstruction(rt.chunk, rt.pc));
  };
}
//#endif
//#endif

export {
  showTree,
  showChunk,
  startTraceExec,
  traceExec,
  startSpan,
  addEvent,
};
