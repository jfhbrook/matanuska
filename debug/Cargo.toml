[workspace]
members = ["crates/debug"]
resolver = "2"
