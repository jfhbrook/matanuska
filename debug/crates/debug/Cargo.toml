[package]
name = "debug"
version = "1.0.0"
description = "Matanuska's debug library"
authors = ["Josh Holbrook"]
license = "MIT"
edition = "2021"
exclude = ["index.node"]

[lib]
crate-type = ["cdylib"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
neon = "1"
tracing = "0.1.41"
tracing-subscriber = "0.3.19"
