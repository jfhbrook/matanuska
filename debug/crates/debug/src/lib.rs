use neon::prelude::*;
use tracing_subscriber;

fn hello(mut cx: FunctionContext) -> JsResult<JsString> {
    Ok(cx.string("hello node"))
}

#[neon::main]
fn main(mut cx: ModuleContext) -> NeonResult<()> {
    tracing_subscriber::fmt::init();
    cx.export_function("hello", hello)?;
    Ok(())
}
