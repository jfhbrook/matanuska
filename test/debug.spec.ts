import { expect, test } from 'vitest';

import { MockTracer } from './helpers/debug';

const TRACER = new MockTracer('main');

test('tracer', async () => {
  TRACER.trace('before span');
  await TRACER.span('span', async () => {
    TRACER.trace('inside span');
  });
  TRACER.trace('after span');
  TRACER.spanSync('sync span', () => {
    TRACER.trace('inside sync span');
  });
  TRACER.trace('after sync span');

  expect(TRACER.messages).toMatchSnapshot();
});
