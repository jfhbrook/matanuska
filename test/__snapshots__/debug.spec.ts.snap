// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tracer 1`] = `
[
  [
    "[32mTRACE[39m <[36mmain[39m>     before span[90m at ///Users/josh/Software/jfhbrook/matanuska/node_modules/@vitest/runner/dist/index.js:533:11[39m",
  ],
  [
    "[32mTRACE[39m <[36mmain[39m>     span[90m at /Users/josh/Software/jfhbrook/matanuska/test/debug.spec.ts:9:21[39m",
  ],
  [
    "[32mTRACE[39m <[36mmain[39m>     [33m|- [39minside span[90m at /Users/josh/Software/jfhbrook/matanuska/test/debug.spec.ts:9:21[39m",
  ],
  [
    "[32mTRACE[39m <[36mmain[39m>     after span[90m at file:///Users/josh/Software/jfhbrook/matanuska/node_modules/@vitest/runner/dist/index.js:986:11[39m",
  ],
  [
    "[32mTRACE[39m <[36mmain[39m>     sync span[90m at /Users/josh/Software/jfhbrook/matanuska/test/debug.spec.ts:13:10[39m",
  ],
  [
    "[32mTRACE[39m <[36mmain[39m>     [33m|- [39minside sync span[90m at /Users/josh/Software/jfhbrook/matanuska/test/debug.spec.ts:13:10[39m",
  ],
  [
    "[32mTRACE[39m <[36mmain[39m>     after sync span[90m at file:///Users/josh/Software/jfhbrook/matanuska/node_modules/@vitest/runner/dist/index.js:986:11[39m",
  ],
]
`;
